A59-audit-logging.md

```diff
@@ -0,0 +1,243 @@
+* Implemented in:
+* Last updated: 2023-03-09
+* Discussion at: 
+
+## Abstract
+
+Support audit logging based on authorization events in gRPC servers in both
+xDS-enabled and gRPC authorization policy based use cases.
+
+## Background
+
+gRPC's authorization support is based on the [RBAC policy][]. In
+xDS-enabled cases, users can configure authorization policies in the control
+plane which applies [RBAC HTTP filters][RBAC filter] to the gRPC servers. When
+xDS is not an option, users can use the [gRPC authorization policy][A43] to
+get similar authorization enforment.
+
+This proposal describes the audit logging APIs that allow users to audit
+specific authorization events in their gRPC servers with the necessary context.
+We aim to offer consistent APIs to support both aforementioned use cases.
+
+[RBAC filter]: https://github.com/envoyproxy/envoy/blob/main/api/envoy/extensions/filters/http/rbac/v3/rbac.proto
+[RBAC policy]: https://github.com/envoyproxy/envoy/blob/main/api/envoy/config/rbac/v3/rbac.proto
+[HttpConnectionManager proto]: https://github.com/envoyproxy/envoy/blob/main/api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto
+
+### Related Proposals: 
+* [gRFC A41: xDS RBAC Support][A41]
+* [gRFC A43: gRPC Authorization API][A43]
+
+[A41]: A41-xds-rbac.md
+[A43]: A43-grpc-authorization-api.md
+
+## Proposal
+
+### Audit Context
+
+The table below lists the metadata we will include in the audit context,
+which will be made available to audit loggers during an audit event.
+
+It is worth noting that audit happens right after the authorization
+decision is made, so the status only reflects that but not the eventual
+one from the server. 
+
+|Fields             |Additional Information                                                             |Example                |
+|-------------------|-----------------------------------------------------------------------------------|-----------------------|
+|RPC Method         |Method the RPC hit.                                                                |"/pkg.service/foo"     |
+|Principal          |Identity the RPC. Currently only available in certificate-based TLS authentication.|"spiffe://foo/user1"   |
+|Timestamp          |Time when the RPC happens.                                                         |1649376211             |
+|Denying Policy Name|Policy name that denied the RPC. Empty if RPC is allowed.                          |"policy_name"          |
+|Status             |gRPC status from the authorization enforcement, not the eventual RPC status.       |OK or PERMISSION_DENIED|
+
+### xDS API Changes
+
+xDS API changes are needed because gRPC authorization is entirely based on [xDS RBAC policy][RBAC policy]. Moreoever, audit logging is a common feature
+that other xDS clients (namely Envoy) should benefit from as well.
+
+We will add an audit condition enum in the [xDS RBAC policy][RBAC policy] as
```

Comment:
You can add a link to the PR in the Envoy repo here.

Reply:
Done.